### PR TITLE
Assignment 4 - counting islands

### PR DESCRIPTION
**Pull request by ElizabethSvit**

### CODE REVIEW

File: ass4.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+void flood_fill_queued(vector<vector<bool> > &table, int y, int x) {
+    deque<pair<int, int>> tile_deque;
+    tile_deque.push_back({y,x});
+    
+    while (!tile_deque.empty()) {
+        pair<int, int> cur_tile = tile_deque.front();
+    
+        tile_deque.pop_front();
+        
+        if (cur_tile.first - 1 >= 0 && table[cur_tile.first - 1][cur_tile.second]) {
```
Comment (ssancho2) [inline ass4.cpp]: A loop on the 4 cardinal directions will be good here (either using two arrays for delta_x & delta_y, or looping over all 8 directions and only using the ones you need).

File: ass4.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        }
+        if (cur_tile.first + 1 < table.size() && table[cur_tile.first + 1][cur_tile.second]) {
+            table[cur_tile.first + 1][cur_tile.second] = false;
+            tile_deque.push_back({cur_tile.first + 1, cur_tile.second});
+        }
+        if (cur_tile.second - 1 >= 0 && table[cur_tile.first][cur_tile.second - 1]) {
+            table[cur_tile.first][cur_tile.second - 1] = false;
+            tile_deque.push_back({cur_tile.first, cur_tile.second - 1});
+        }
+        if (cur_tile.second + 1 < table[0].size() && table[cur_tile.first][cur_tile.second + 1]) {
```
Comment (ssancho2) [inline ass4.cpp]: Not a big deal, but what will happen if the first row on the table is bigger than the rest?

File: ass4.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+    }
+    table[y][x] = false;
+    
+    flood_fill_recursive(table, y - 1, x);
+    flood_fill_recursive(table, y + 1, x);
+    flood_fill_recursive(table, y, x - 1);
+    flood_fill_recursive(table, y, x + 1);
+}
+
+// Second variant of algorithm using data structure deque for storing islands
```
Comment (ssancho2) [inline ass4.cpp]: Good that you have both variants, when will you use each? :)

File: ass4.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
+#include <iostream>
+#include <vector>
+#include <deque>
+
+using namespace std;
+
+// First variant of algorithm using recursion
+
+void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
```
Comment (ssancho2) [inline ass4.cpp]: A pointer is better as a parameter you want to modify, make the calling code more explicit, compare:
`flood_fill_recursive(table, y, x)
`vs
`flood_fill_recursive(&table, y, x)`
In the second is obvious that the table will be modified.

**Review by ssancho2 — COMMENTED** [review]
Looks good! Did you think of the other solution (re: hint 2 on the assignment) ?

File: ass4.cpp (hunk 3 above)
Comment (ElizabethSvit) [reply ass4.cpp]: It was just for practising both recursive and non-recursive algorithms :) I left only recursive for now. Also added the second algorithm Union-Find from Hint 2.

File: ass4.cpp [hunk 5]
```diff
@@ -1,12 +1,11 @@
...
 #include <vector>
-#include <deque>
 
 using namespace std;
 
-// First variant of algorithm using recursion
+// Recursive "Flood fill" algorithm
 
-void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
+void flood_fill_recursive(vector<vector<bool> > table, int y, int x) {
```
Comment (ssancho2) [inline ass4.cpp]: This way you are making a copy of the whole table for every call, and you are not modifying the original table that was passed to you. Didn't the tests break?

File: ass4.cpp [hunk 6]
```diff
@@ -22,49 +21,112 @@ void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
...
+        rank.resize(n, 0);
+        for (int i = 0; i < n; ++i) {
+            parent.push_back(i);
+        }
+    }
     
-        tile_deque.pop_front();
+    int find(int x) {
+        if (parent[x] != x) {
+            return find(parent[x]);
```
Comment (ssancho2) [inline ass4.cpp]: It's a very common optimization to set `parent[x] = find(parent[x])` this way the tree gets shallower and it's faster the next time around.

File: ass4.cpp [hunk 7]
```diff
@@ -22,49 +21,112 @@ void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
...
+        }
+    }
+    
+    Disjoint_sets islands = Disjoint_sets(table.size() * max_width);
+    vector<pair<int, int>> delta_x_y = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
+
+    for (int y = 0; y < table.size(); ++y) {
+        for (int x = 0; x < table[y].size(); ++x) {
+            if (table[y][x]) {
+                for (auto delta : delta_x_y) {
```
Comment (ssancho2) [inline ass4.cpp:98]: Actually, you only need to join with 2, up and left (or up and right or similar :) ), that may make the code simpler.

File: ass4.cpp [hunk 8]
```diff
@@ -22,49 +21,112 @@ void flood_fill_recursive(vector<vector<bool> > &table, int y, int x) {
...
+    for (int y = 0; y < table.size(); ++y) {
+        for (int x = 0; x < table[y].size(); ++x) {
+            if (max_width < table[y].size()) {
+                max_width = table[y].size();
+            }
+        }
+    }
+    
+    Disjoint_sets islands = Disjoint_sets(table.size() * max_width);
+    vector<pair<int, int>> delta_x_y = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
```
Comment (ssancho2) [inline ass4.cpp:93]: Nice

File: ass4.cpp [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+            rank[y_ancestor] = rank[y_ancestor] + 1;
+        }
+    }
+};
+
+int count_islands(vector<vector<bool> > table) {
+    int max_width = 0;
+    
+    for (int y = 0; y < table.size(); ++y) {
+        for (int x = 0; x < table[y].size(); ++x) {
```
Comment (ssancho2) [inline ass4.cpp:85]: No need for this loop.

File: ass4.cpp [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+                        && table[y + delta.first][x + delta.second]) {
+                        islands.func_union(y + delta.first * table[y + delta.first].size() + x + delta.second,
+                                           (y + delta.first) * table[y + delta.first].size() + x + delta.second);
+                    }
+                }
+            }
+        }
+    }
+
+    vector<int> components(table.size() * max_width);
```
Comment (ssancho2) [inline ass4.cpp:110]: Easier to use a set, add all values returned by find on the islands and see the size at the end.

File: ass4.cpp [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+            }
+        }
+    }
+    return counter;
+}
+
+
+int main(int argc, const char * argv[]) {
+    vector<vector<bool> > table = {{0,0,1}, {1,0,0}, {1,1,0}};
+    cout << count_islands(table) << endl;
```
Comment (ssancho2) [inline ass4.cpp:132]: Some more test will be nice :)

**Review by ssancho2 — APPROVED** [review]
Some comments, but fell free to move to the next assignment if you don't have time.

File: ass4.cpp (hunk 5 above)
Comment (ElizabethSvit) [reply ass4.cpp]: Yes, I removed it accidentially and tested only second algorithm after that.

**Pull request merged**